Blog Post: OpenSearch plugin zips now in Maven repo (#930)
opensearch-project/project-website: _posts/2022-07-24-opensearch-plugin-zips-now-in-maven-repo.markdown

---
layout: post
title:  "OpenSearch plugin zips now in Maven repo"
authors: 
  - prudhvigodithi
date: 2022-07-24 01:01:01 -0700
categories: 
  - technical-post
twittercard:

>   description: "Details on how to consume OpenSearch plugin zips from maven repo and the process involved in shipping them to the maven repo."

Reviewer: Capitalize both instances of "Maven". Additionally, should it be "Details on how to consume OpenSearch plugin zips from a Maven repo and the process involved in shipping them to a Maven repo."?

**Figure 1**: Workflow that ship generated plugin zips to maven

![Figure 1: Workflow that ship zips to maven]({{ site.baseurl }}/assets/media/blog-images/2022-07-24-opensearch-plugin-zips-now-in-maven-repo/figure1.png){: .img-fluid }

### Benefits of plugin zips in maven:

* The plugin dependencies can now be fetched using the right maven coordinates from maven repo and a user need not use a zip of some random version built on a developer desktop.
* The development dependent `SNAPSHOT` version zips can be directly fetched from maven using the dependency mechanism.
* Not required to check in the zips manually to any `src/` folders, as zips can now be fetched using the right `grounID`, `artifactID` and `version`.
* The published zips to central maven repo are signed with `.asc`, `.md5`, `.sha1`, `.sha256`, `.sha512` extensions.
* The tests and CI workflows can directly run against the zips fetched from maven repo, instead of requiring a manual download and checking into the project.

### Consume plugin zips from maven:

> **Using mvn cli:**

Reviewer: Should this be "Using the Maven CLI"? Remove the colon (:) from the heading.

> Consume from Central Maven repo:

Reviewer: Change to "Consume from the central Maven repo:".

```
mvn org.apache.maven.plugins:maven-dependency-plugin:2.1:get -DrepoUrl=https://repo1.maven.org/maven2 -Dartifact=org.opensearch.plugin:opensearch-job-scheduler:2.1.0.0:zip
```

> Consume from Snapshot Maven repo:

Reviewer: Add "the" before "Snapshot".

```
mvn org.apache.maven.plugins:maven-dependency-plugin:2.1:get -DrepoUrl=https://aws.oss.sonatype.org/content/repositories/snapshots -Dartifact=org.opensearch.plugin:opensearch-job-scheduler:2.1.0.0-SNAPSHOT:zip
```

> **Gradle Project: using build.gradle file**

Reviewer: Change to "Gradle project: Using the build.gradle file".

```
dependencies {
    classpath "org.opensearch.plugin:opensearch-job-scheduler:2.1.0.0@zip"
    classpath "org.opensearch.plugin:opensearch-knn:2.1.0.0@zip"
}
```

> **Maven Project: using pom.xml file**

Reviewer: Change to "Maven project: Using the pom.xml file".

---
layout: post
title:  "OpenSearch plugin zips now in Maven repo"
authors: 
  - prudhvigodithi
date: 2022-07-24 01:01:01 -0700
categories: 
  - technical-post
twittercard:
  description: "Details on how to consume OpenSearch plugin zips from maven repo and the process involved in shipping them to the maven repo."
---

> 

Reviewer 2: I deleted the Summary section and used that description as the Intro.

```suggestion
Starting with the release of OpenSearch `2.1.0`, OpenSearch plugin zips are now signed and published to a central Apache Maven [repo](https://repo1.maven.org/maven2/org/opensearch/plugin/).  Using the [Release zips Maven Repo](https://repo1.maven.org/maven2/org/opensearch/plugin/) and [Snapshot zips Maven Repo](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/) URL’s, OpenSearch plugin zips can now be consumed  as a dependency to build other plugins or fetched as standalone components for your OpenSearch cluster.
```

Author: Sentence Reformation:

Starting with the release of OpenSearch `2.1.0`, OpenSearch plugin zips are now signed and published to a central Apache Maven [repo](https://repo1.maven.org/maven2/org/opensearch/plugin/). Using the [Release zips](https://repo1.maven.org/maven2/org/opensearch/plugin/) and [Snapshot zips](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/) Maven Repo URL’s, OpenSearch plugin zips can now be consumed  as a dependency to build other plugins or fetched as standalone components for your OpenSearch cluster.

> ### Motivation and Current Challenges:

Reviewer 2: ```suggestion
## Motivation
```

> The plugins zips that are used as dependencies cannot be downloaded dynamically during runtime, as they are not part of version controlled maven system. Only the respective Java jars are available to download using maven coordinates. This forces a user to now to use a dependency plugin zip of some random version built on a developer desktop, which is not reliable, also the tests executed as part of this build zip cannot be accurate, so eventually for few plugins the local build zip is forced to check in to the plugin repo to facilitate the plugin availability for dependency, this is one of the challenge faced by the community. Coming to the download part of plugins as separate isolated components via a cached mechanism like maven is not possible, with this challenge users had to  pre-bake the random built zip to the project.

Reviewer 2: ```suggestion
Before OpenSearch 2.1, plugin zips used as dependencies could not be downloaded dynamically during runtime, because plugin zips were not a part of the version-controlled Maven system. The only mechanism for plugin downloads was OpenSearch respective Java jars fetched through Maven coordinates. The Java jars forced users who wanted more control over their OpenSearch plugin configuration to use a random plugin version built on a developer desktop, instead of a more reliable version-controlled plugin. 
```

Author: Can we re-form the sentence to match the meaning as follows

Before OpenSearch `2.1`, plugin zips used as dependencies could not be downloaded dynamically during runtime, because plugin zips were not a part of the version-controlled Maven system. Only the respective Java jars are available to download using maven coordinates. This forces a user to now to use a dependency plugin zip of some random version built on a developer desktop, instead of a more reliable version-controlled plugin.

> Starting with release `2.1.0` OpenSearch plugin zips are now signed and published to central maven repo, with `groupID` as [org.opensearch.plugin](https://repo1.maven.org/maven2/org/opensearch/plugin/). These zips can now be fetched as individual components by directly downloading using maven coordinates or by clickstream from central maven repo which can be later cached to the desired user local maven repo. The snapshot version of the zips can be fetched from nexus maven repo with the same `groupID` as [org.opensearch.plugin](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/).

Reviewer 2: ```suggestion
Furthermore, tests executed against as part ofthe OpenSearch build process from zip were not accurate, as each local build zip had to check against the plugin repo in order to ensure that plugin's availability. These restrictions were challenging to our community, since those wishing to use plugin zips as separate isolated components via a cached mechanism proved to be impossible. 
```

Author: This looks really good, small change though 

Furthermore, to facilitate the plugin availability as dependency, the plugin zip's are forced to check in to the plugin repo and so tests executed as part of the OpenSearch build process not accurate. These restrictions were challenging to our community, since those wishing to use plugin zips as separate isolated components via a cached mechanism proved to be impossible.

Reviewer: Add "a" between "as" and "dependency" ("as a dependency"). Change "zip's" to "zips" (no apostrophe). Change to "...to the plugin repo, so tests...". Remove the comma after "community". Change "since those wishing to use" to "since using".

> ### Maven publish task with custom gradle plugin.

Reviewer 2: I also changed the order of the "Benefits" and "Gradle" sections. 

```suggestion
## Benefits of plugin zips in Maven
```

> The publishing of zips to central maven repo had been possible with custom gradle plugin `opensearch.pluginzip`.

Reviewer 2: ```suggestion
With Maven, plugin zips can now be retrieved by:

- Downloading each plugin directly using their respective Maven coordinates 
- Using clickstream from the central Maven repo, which can be cached later to a local Maven repo.
- Fetching the development `Snapshot` version with same Mavin `groupID` as [org.opensearch.plugin](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/).

```

Author: Just some additional change to last point

- Fetching the development `SNAPSHOT` version with same Maven `groupID` as [org.opensearch.plugin](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/).

Reviewer: Add a terminating period to the first list item. In the second list item, should it be "a clickstream"? In the third list item, change "with same Mavin" to "with the same Maven".

> Starting with release `2.1.0` all OpenSearch gradle supported plugins will have custom gradle plugin `opensearch.pluginzip` applied to the gradle project, this will create a new task `publishPluginZipPublicationToZipStagingRepository` which does all the heavy lifting in identifying the distribution plugin zip, setting the maven coordinates, generating the POM file, updating the user passed POM fields and publishing it to maven repo. More details and inner workings of the plugin can found at [opensearch-plugins](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip).

Reviewer 2: ```suggestion
Using OpenSearch plugin zips through Maven offers the following benefits:

- Plugins zip in the central Maven repo is already signed with `.asc`, `.md5`, `.sha1`, `.sha256`, `.sha512` extensions.
- Users are no longer required to to check in zips to any `src/` files, since zips can be fetched with the right `groupID`, `artifactID` and `version`.
- Tests and CI workflows can directly run against the Maven repo, instead of requiring a manual download. 
```

Author: - Tests and CI workflows can directly run against the zips from Maven repo, instead of requiring a manual download.

Reviewer: In the first list item, change to "Plugin zips in the central Maven repo are already signed with...". Add "and" after the last comma. In the second list item, remove the comma after "files" and add a comma after "'artifactID'". In the third list item, change to "Tests and CI workflows can run directly against the Maven repo instead of requiring a manual download."

> **Figure 1**: Workflow that ship generated plugin zips to maven

Reviewer 2: ```suggestion

```

> ![Figure 1: Workflow that ship zips to maven]({{ site.baseurl }}/assets/media/blog-images/2022-07-24-opensearch-plugin-zips-now-in-maven-repo/figure1.png){: .img-fluid }

Reviewer 2: ```suggestion
```

> ### Benefits of plugin zips in maven:

Reviewer 2: ```suggestion
### Maven zip publication and gradle
```

Author: Heading change as `Maven zip publication with Gradle`

> * The plugin dependencies can now be fetched using the right maven coordinates from maven repo and a user need not use a zip of some random version built on a developer desktop.

Reviewer 2: ```suggestion
## Maven zip publication and gradle

OpenSearch publishes plugin zips using a custom gradle plugin `opensearch.pluginzip`. With OpenSearch 2.1, all OpenSearch gradle supported plugins create a new task, `publishPluginZipPublicationToZipStagingRepository`. The task performs all the heavy lifting for users by:

- Identifying the distribution plugin zip
- Setting the Maven coordinates
- Generating the POM file
- Updating the user-generated PM fields
- Publishing the task to your Maven repo

You can find more details about the inner workings of OpenSearch plugins in the [opensearch-plugins repo](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip).

**Figure**: Workflow that ship generated plugin zips to maven

![Figure: Workflow that ship zips to maven]({{ site.baseurl }}/assets/media/blog-images/2022-07-24-opensearch-plugin-zips-now-in-maven-repo/figure1.png){: .img-fluid }
```

Author: Heading change as `Maven zip publication with Gradle`

Add `2.1` in code quotes.

A small adjustment in 4th point as follows:
- Updating the user passed POM fields

Reviewer: Capitalize all instances of "Gradle". Line 33: Add a comma after "OpenSearch publishes plugin zips using a custom Gradle plugin,". Change to "Gradle-supported". Add terminating periods to all list items. The fourth list item should read "Updating the user-generated POM fields." The Figure caption should read "Workflow that ships generated plugin zips to Maven".

> * The development dependent `SNAPSHOT` version zips can be directly fetched from maven using the dependency mechanism.

Reviewer 2: ```suggestion

```

> * Not required to check in the zips manually to any `src/` folders, as zips can now be fetched using the right `grounID`, `artifactID` and `version`.

Reviewer 2: ```suggestion

```

> * The published zips to central maven repo are signed with `.asc`, `.md5`, `.sha1`, `.sha256`, `.sha512` extensions.

Reviewer 2: ```suggestion

```

> * The tests and CI workflows can directly run against the zips fetched from maven repo, instead of requiring a manual download and checking into the project.

Reviewer 2: ```suggestion

```

> ### Consume plugin zips from maven:

Reviewer 2: ```suggestion
## Consume plugin zips from Maven
```

> 

Reviewer 2: ```suggestion
You can fetch and use OpenSearch plugin zips from Maven in three ways:
```

**Using mvn cli:**

Consume from Central Maven repo:
```

> mvn dependency:get -DgroupId=org.opensearch.plugin -DartifactId=opensearch-job-scheduler -Dversion=2.1.0.0 -Dpackaging=zip -DremoteRepositories=central::default::https://repo.maven.apache.org/maven2,myrepo::::http://myrepo.com/maven2

Author: Small change with the command (To make it even more clear)

Small change with the command (To make it even more clear)
```
mvn dependency:get -DgroupId=org.opensearch.plugin -DartifactId=opensearch-job-scheduler -Dversion=2.1.0.0 -Dpackaging=zip -DremoteRepositories=https://repo1.maven.org/maven2
```

```
Consume from Snapshot Maven repo:
```
Snapshot Maven repo
mvn dependency:get -DgroupId=org.opensearch.plugin -DartifactId=opensearch-job-scheduler -Dversion=2.1.0.0-SNAPSHOT -Dpackaging=zip -DremoteRepositories=https://aws.oss.sonatype.org/content/repositories/snapshots/
```

**Gradle Project: using build.gradle file**

```
dependencies {
    classpath "org.opensearch.plugin:opensearch-job-scheduler:2.1.0.0@zip"
    classpath "org.opensearch.plugin:opensearch-knn:2.1.0.0@zip"
}
```

**Maven Project: using pom.xml file**

```
<dependencies>
        <dependency>
            <groupId>org.opensearch.plugin</groupId>
            <artifactId>opensearch-job-scheduler</artifactId>
            <version>2.1.0.0</version>
            <packaging>zip</packaging>
        </dependency>
</dependencies>
```

> ### Summary:

Reviewer 2: ```suggestion

```

Author: Small change with the command

Consume from Central Maven repo:
```
mvn dependency:get -DgroupId=org.opensearch.plugin -DartifactId=opensearch-job-scheduler -Dversion=2.1.0.0 -Dpackaging=zip -DremoteRepositories=https://repo1.maven.org/maven2
```

> Using the maven [Release zips Maven Repo](https://repo1.maven.org/maven2/org/opensearch/plugin/) and [Snapshot zips Maven Repo](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/) URL’s, the plugin zips can be consumed and used as dependency to build other plugins and can also be fetched as standalone components to install them to the OpenSearch cluster.

Reviewer 2: ```suggestion

```

---
layout: post
title:  "OpenSearch plugin zips now in Maven repo"
authors: 
  - prudhvigodithi
date: 2022-07-24 01:01:01 -0700
categories: 
  - technical-post
twittercard:
  description: "Details on how to consume OpenSearch plugin zips from a Maven repo and the process involved in shipping them to a Maven repo."
---

Starting with the release of OpenSearch `2.1.0`, OpenSearch plugin zips are now signed and published to a central Apache Maven [repo](https://repo1.maven.org/maven2/org/opensearch/plugin/).  Using the [Release zips](https://repo1.maven.org/maven2/org/opensearch/plugin/) and [Snapshot zips](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/) Maven Repo URLs, OpenSearch plugin zips can now be consumed  as a dependency to build other plugins or fetched as standalone components for your OpenSearch cluster. 

## Motivation 

> Before OpenSearch 2.1, plugin zips used as dependencies could not be downloaded dynamically during runtime because plugin zips were not a part of the version-controlled Maven system. The only mechanism for plugin downloads was each plugins' respective Java jars through Maven coordinates. This system forced users who wanted more control over their OpenSearch plugin configuration to use a dependency plugin zip built on a developer desktop instead of a more reliable version-controlled plugin. 

Reviewer: Change "plugins'" to "plugin's".

> Furthermore, to facilitate the plugin availability as a dependency, tests executed against as part the OpenSearch build process from zip were not accurate, as each local build zip had to find the plugin repo in order to ensure that plugin's availability. These restrictions were challenging to our community since using plugin zips as separate isolated components via a cached mechanism proved to be impossible. 

Reviewer: Currently reads "...tests executed against as part the...". Please correct. Change "since" to "because".

## Benefits of Maven

With Maven, plugin zips can now be retrieved by:

> - Downloading each plugin directly using their respective Maven coordinates .

Reviewer: Remove the space after the period.

- Using clickstream from the central Maven repo, which can be cached later to a local Maven repo.

> - Fetching the development `SNAPSHOT` version with same Maven `groupID` as [org.opensearch.plugin](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/).

Reviewer: Add "the" before "same".

Using OpenSearch plugin zips through Maven offers the following benefits:

- Plugins zip in the central Maven repo are already signed with `.asc`, `.md5`, `.sha1`, `.sha256`, and `.sha512` extensions.

> - Users are no longer required to to check in zips to any `src/` files since zips can be fetched with the right `groupID`, `artifactID`, and `version`.

Reviewer: Change "since" to "because".

> - Tests and CI workflows can directly run against zips from the Maven repo instead of requiring a manual download. 

Reviewer: Define "CI" as "continuous integration (CI)".

## Maven zip publication with Gradle

OpenSearch publishes plugin zips using a custom Gradle plugin, `opensearch.pluginzip`. With OpenSearch 2.1, all OpenSearch Gradle-supported plugins create a new task, `publishPluginZipPublicationToZipStagingRepository`. The task performs all the heavy lifting for users by:

- Identifying the distribution plugin zip.
- Setting the Maven coordinates.
- Generating the POM file.
- Updating with the user-generated POM fields.
- Publishing the zip to your Maven repo.

You can find more details about the inner workings of OpenSearch plugins in the [opensearch-plugins repo](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip).

**Figure 1**: Workflow that ships generated plugin zips to Maven

![Figure 1: Workflow that ship zips to maven]({{ site.baseurl }}/assets/media/blog-images/2022-07-24-opensearch-plugin-zips-now-in-maven-repo/figure1.png){: .img-fluid }**Figure 1**: Workflow that ships generated plugin zips to Maven

## Consume plugin in zips

> You can fetch plugin in zips in three different ways:

Reviewer: Should "plugin" be "plugins" in both instances, or should there be an article in front of "plugin" in both instances? Change the colon to a period.

**Using the Maven CLI**

> Consume from central Maven repo:

Reviewer: Add "the" before "central".
